src/extensions/server/mod.rs

```diff
@@ -90,40 +96,79 @@ impl SubwayServerBuilder {
 
     pub async fn build<Fut: Future<Output = anyhow::Result<RpcModule<()>>>>(
         &self,
-        rate_limit: Option<RateLimit>,
-        builder: impl FnOnce() -> Fut,
+        rate_limit_builder: Option<Arc<RateLimitBuilder>>,
+        rpc_module_builder: impl FnOnce() -> Fut,
     ) -> anyhow::Result<(SocketAddr, ServerHandle)> {
-        let rpc_middleware = RpcServiceBuilder::new().option_layer(rate_limit);
-
-        let service_builder = tower::ServiceBuilder::new()
-            .layer(cors_layer(self.config.cors.clone()).expect("Invalid CORS config"))
-            .layer(
-                ProxyGetRequestLayer::new(
-                    self.config
-                        .http_methods
-                        .iter()
-                        .map(|m| ProxyGetRequestMethod {
-                            path: m.path.clone(),
-                            method: m.method.clone(),
-                        })
-                        .collect(),
-                )
-                .expect("Invalid health config"),
-            );
-
-        let server = ServerBuilder::default()
-            .set_rpc_middleware(rpc_middleware)
-            .set_http_middleware(service_builder)
-            .max_connections(self.config.max_connections)
-            .set_id_provider(RandomStringIdProvider::new(16))
-            .build((self.config.listen_address.as_str(), self.config.port))
-            .await?;
-
-        let module = builder().await?;
-
-        let addr = server.local_addr()?;
-        let server = server.start(module);
-
-        Ok((addr, server))
+        let config = self.config.clone();
+
+        let (stop_handle, server_handle) = stop_channel();
+        let handle = stop_handle.clone();
+        let rpc_module = rpc_module_builder().await?;
+
+        // make_service handle each connection
+        let make_service = make_service_fn(move |socket: &AddrStream| {
+            let remote_ip = socket.remote_addr().ip().to_string();
```

Comment:
```suggestion
            let socket_ip = socket.remote_addr().ip().to_string();
```

to make everything bit more easier to understand